base-server.d.ts

```diff
@@ -0,0 +1,704 @@
+  id: string
+  time: number
+  subprotocol: LoguxBaseServerOptions['subprotocol']
+  server: string
+  reasons: string[]
+  channels?: string[]
+  clients?: string[]
+  status?: 'waiting' | 'add' | 'clean' | 'processed' | 'subscribed'
+}
+
+/**
+ * Logux base action
+ */
+export type LoguxBaseAction = {
+  type: string
+  id?: string
+  channel?: string
+  since?: {
+    id: string
+    time: number
+  }
+}
+
+/**
+ * BaseServer options.
+ */
+export type LoguxBaseServerOptions = {
+  /**
+   * Server current application subprotocol version in SemVer format.
+   */
+  subprotocol: string
+
+  /**
+   * npm’s version requirements for client subprotocol version.
+   */
+  supports: string
+
+  /**
+   * Application root to load files and show errors.
+   * Default is `process.cwd()`.
+   */
+  root?: string
+
+  /**
+   * Timeout in milliseconds to disconnect connection.
+   * Default is `20000`.
+   */
+  timeout?: number
+
+  /**
+   * Milliseconds since last message to test connection by sending ping.
+   * Default is `10000`.
+   */
+  ping?: number
+
+  /**
+   * URL to PHP, Ruby on Rails, or other backend to process actions and
+   * authentication.
+   */
+  backend?: string
+
+  /**
+   * URL to Redis for Logux Server Pro scaling.
+   */
+  redis?: string
+
+  /**
+   * Host to bind HTTP server to control Logux server.
+   * Default is `127.0.0.1`.
+   */
+  controlHost?: string
+
+  /**
+   * Port to control the server. Default is `31338`.
+   */
+  controlPort?: number
+
+  /**
+   * Password to control the server.
+   */
+  controlPassword?: string
+
+  /**
+   * Store to save log. Will be {@link @logux/core:MemoryStore}, by default.
+   */
+  store?: Store
+
+  /**
+   * Test time to test server.
+   */
+  time?: TestTime
+
+  /**
+   * Custom random ID to be used in node ID.
+   */
+  id?: string
+
+  /**
+   * Development or production server mode. By default,
+   * it will be taken from `NODE_ENV` environment variable.
+   * On empty `NODE_ENV` it will be `'development'`.
+   */
+  env?: 'production' | 'development'
+
+  /**
+   * Process ID, to display in reporter.
+   */
+  pid?: number
+
+  /**
+   * HTTP server to connect WebSocket server to it. Same as in `ws.Server`.
+   */
+  server?: HTTPServer
+
+  /**
+   * Port to bind server. It will create HTTP server manually to connect
+   * WebSocket server to it. Default is `31337`.
+   */
+  port?: number
+
+  /**
+   * IP-address to bind server. Default is `127.0.0.1`.
+   */
+  host?: string
+
+  /**
+   * SSL key or path to it. Path could be relative from server root.
+   * It is required in production mode, because WSS is highly recommended.
+   */
+  key?: string
+
+  /**
+   * SSL certificate or path to it. Path could be relative from server
+   * root. It is required in production mode, because WSS is highly
+   * recommended.
+   */
+  cert?: string
+
+  /**
+   * Function to show current server status.
+   */
+  reporter?: (event: string, payload: Object) => void
```

Comment:
Can we express `payload` according to `event` type like we did in Logux Client?
https://github.com/logux/client/blob/master/cross-tab-client/index.d.ts#L89-L92